File: backend-dotnet/src/Veha.Api/Services/RiskService.cs
using Microsoft.EntityFrameworkCore;
using Veha.Api.Dtos;
using Veha.Domain.Common;
using Veha.Domain.Entities;
using Veha.Domain.Enums;
using Veha.Domain.Rules;
using Veha.Infrastructure;

namespace Veha.Api.Services;

/// <summary>Реестр рисков: CRUD с аудитом, пересчёт score, портфель (матрица 3×3).
/// Порт services/risk_service.py. Score = probability×impact денормализован.</summary>
public class RiskService(
    VehaDbContext db, AuditService audit, NotificationService notifier, ProjectAccessService access)
{
    private static readonly RiskStatus[] Active = [RiskStatus.Open, RiskStatus.Mitigating];
    private static readonly string[] InternalRoles = ["admin", "director", "pm", "presale", "engineer", "finance"];

    public async Task<List<RiskOutDto>> ListForProjectAsync(Guid projectId, RiskStatus? status, CancellationToken ct)
    {
        var q = db.Risks.Where(r => r.ProjectId == projectId);
        if (status is not null) q = q.Where(r => r.Status == status.Value);
        return (await q.OrderByDescending(r => r.Score).ThenByDescending(r => r.CreatedAt).ToListAsync(ct))
            .Select(ToOut).ToList();
    }

    public async Task<Risk> GetAsync(Guid riskId, CancellationToken ct)
        => await db.Risks.FirstOrDefaultAsync(r => r.Id == riskId, ct)
           ?? throw new NotFoundException("Риск не найден");

    public async Task<RiskOutDto> CreateAsync(Guid projectId, RiskCreateDto dto, Guid actorId, CancellationToken ct)
    {
        var score = RiskRules.ComputeScore(dto.Probability, dto.Impact);
        var risk = new Risk
        {
            ProjectId = projectId, CreatedBy = actorId, Score = score,
            Title = dto.Title, Description = dto.Description, Category = dto.Category,
            Probability = dto.Probability, Impact = dto.Impact,
            ResponseStrategy = dto.ResponseStrategy, MitigationPlan = dto.MitigationPlan,
            OwnerId = dto.OwnerId, DueDate = dto.DueDate,
        };
        db.Risks.Add(risk);
        await audit.RecordAsync("Risk", risk.Id, AuditAction.Create, actorId,
            new { title = risk.Title, score, level = RiskRules.RiskLevel(score) }, ct);
        if (risk.OwnerId is not null && risk.OwnerId != actorId)
            await NotifyOwnerAsync(risk, ct);
        return ToOut(risk);
    }

    public async Task<RiskOutDto> UpdateAsync(Guid riskId, RiskUpdateDto dto, Guid actorId, CancellationToken ct)
    {
        var risk = await GetAsync(riskId, ct);
        var prevOwner = risk.OwnerId;
        var diff = new Dictionary<string, object>();
        void Track(string k, object? oldV, object? newV) => diff[k] = new { from = oldV, to = newV };

        if (dto.Title is not null && dto.Title != risk.Title) { Track("title", risk.Title, dto.Title); risk.Title = dto.Title; }
        if (dto.Description is not null && dto.Description != risk.Description) { Track("description", risk.Description, dto.Description); risk.Description = dto.Description; }
        if (dto.Category is not null && dto.Category.Value != risk.Category) { Track("category", risk.Category, dto.Category.Value); risk.Category = dto.Category.Value; }
        var probImpactChanged = false;
        if (dto.Probability is not null && dto.Probability.Value != risk.Probability) { Track("probability", risk.Probability, dto.Probability.Value); risk.Probability = dto.Probability.Value; probImpactChanged = true; }
        if (dto.Impact is not null && dto.Impact.Value != risk.Impact) { Track("impact", risk.Impact, dto.Impact.Value); risk.Impact = dto.Impact.Value; probImpactChanged = true; }
        var statusProvided = dto.Status is not null;
        if (dto.Status is not null && dto.Status.Value != risk.Status) { Track("status", risk.Status, dto.Status.Value); risk.Status = dto.Status.Value; }
        if (dto.ResponseStrategy is not null && dto.ResponseStrategy.Value != risk.ResponseStrategy) { Track("response_strategy", risk.ResponseStrategy, dto.ResponseStrategy.Value); risk.ResponseStrategy = dto.ResponseStrategy.Value; }
        if (dto.MitigationPlan is not null && dto.MitigationPlan != risk.MitigationPlan) { Track("mitigation_plan", risk.MitigationPlan, dto.MitigationPlan); risk.MitigationPlan = dto.MitigationPlan; }
        if (dto.OwnerId is not null && dto.OwnerId != risk.OwnerId) { Track("owner_id", risk.OwnerId, dto.OwnerId); risk.OwnerId = dto.OwnerId; }
        if (dto.DueDate is not null && dto.DueDate != risk.DueDate) { Track("due_date", risk.DueDate, dto.DueDate); risk.DueDate = dto.DueDate; }

        if (probImpactChanged)
        {
            var newScore = RiskRules.ComputeScore(risk.Probability, risk.Impact);
            if (newScore != risk.Score) { Track("score", risk.Score, newScore); risk.Score = newScore; }
        }

        if (statusProvided)
        {
            if (risk.Status == RiskStatus.Closed && risk.ClosedAt is null) risk.ClosedAt = DateTimeOffset.UtcNow;
            else if (risk.Status != RiskStatus.Closed) risk.ClosedAt = null;
        }

        if (diff.Count > 0)
            await audit.RecordAsync("Risk", risk.Id, AuditAction.Update, actorId, diff, ct);
        else
            await db.SaveChangesAsync(ct);

        if (risk.OwnerId is not null && risk.OwnerId != prevOwner && risk.OwnerId != actorId)
            await NotifyOwnerAsync(risk, ct);
        return ToOut(risk);
    }

    public async Task DeleteAsync(Guid riskId, Guid actorId, CancellationToken ct)
    {
        var risk = await GetAsync(riskId, ct);
        risk.DeletedAt = DateTimeOffset.UtcNow;
        await audit.RecordAsync("Risk", risk.Id, AuditAction.Delete, actorId, new { title = risk.Title }, ct);
    }

    private async Task NotifyOwnerAsync(Risk risk, CancellationToken ct)
    {
        if (risk.OwnerId is null) return;
        var owner = await db.Users.FirstOrDefaultAsync(u => u.Id == risk.OwnerId, ct);
        if (owner is null || !InternalRoles.Any(owner.Roles.Contains)) return;
        await notifier.NotifyAsync(risk.OwnerId.Value, "risk_assigned",
            "Вы назначены ответственным за риск", risk.Title, $"/projects/{risk.ProjectId}", ct);
    }

    // ---------- Портфель рисков (set-based) ----------
    public async Task<RiskPortfolioResponseDto> PortfolioAsync(User user, CancellationToken ct)
    {
        var projects = await access.AccessibleProjectsAsync(user, null, 500, ct);
        var byId = projects.ToDictionary(p => p.Id);
        var ids = byId.Keys.ToList();
        if (ids.Count == 0) return EmptyPortfolio();

        var rows = await db.Risks
            .Where(r => ids.Contains(r.ProjectId) && Active.Contains(r.Status))
            .Select(r => new { r.ProjectId, r.Probability, r.Impact, r.Score, r.Category })
            .ToListAsync(ct);

        var matrix = new Dictionary<(int, int), int>();
        var byCategory = new Dictionary<RiskCategory, int>();
        var perProject = new Dictionary<Guid, (int Active, int High, int Top)>();
        var totalHigh = 0;
        foreach (var r in rows)
        {
            matrix[(r.Probability, r.Impact)] = matrix.GetValueOrDefault((r.Probability, r.Impact)) + 1;
            byCategory[r.Category] = byCategory.GetValueOrDefault(r.Category) + 1;
            var agg = perProject.TryGetValue(r.ProjectId, out var e) ? e : (Active: 0, High: 0, Top: 0);
            agg.Active++;
            agg.Top = Math.Max(agg.Top, r.Score);
            if (r.Score >= RiskRules.HighThreshold) { agg.High++; totalHigh++; }
            perProject[r.ProjectId] = agg;
        }

        var matrixCells = (from p in Enumerable.Range(1, 3)
                           from i in Enumerable.Range(1, 3)
                           select new RiskMatrixCellDto(p, i, p * i, RiskRules.RiskLevel(p * i), matrix.GetValueOrDefault((p, i)))).ToList();

        var projectRows = perProject
            .Where(kv => byId.ContainsKey(kv.Key))
            .Select(kv =>
            {
                var p = byId[kv.Key];
                return new ProjectRiskRowDto(p.Id, p.Code, p.Name, p.Stage, p.Status, p.ManagerId,
                    kv.Value.Active, kv.Value.High, kv.Value.Top);
            })
            .OrderByDescending(r => r.TopScore).ThenByDescending(r => r.ActiveCount)
            .ToList();

        var categoryCounts = byCategory
            .OrderByDescending(kv => kv.Value)
            .Select(kv => new RiskCategoryCountDto(kv.Key, kv.Value)).ToList();

        return new RiskPortfolioResponseDto(
            projectRows.Count, rows.Count, totalHigh, matrixCells, categoryCounts, projectRows);
    }

    private static RiskPortfolioResponseDto EmptyPortfolio()
    {
        var cells = (from p in Enumerable.Range(1, 3)
                     from i in Enumerable.Range(1, 3)
                     select new RiskMatrixCellDto(p, i, p * i, RiskRules.RiskLevel(p * i), 0)).ToList();
        return new RiskPortfolioResponseDto(0, 0, 0, cells, [], []);
    }

    private static RiskOutDto ToOut(Risk r) => new(
        r.Id, r.ProjectId, r.Title, r.Description, r.Category, r.Probability, r.Impact, r.Score,
        r.Status, r.ResponseStrategy, r.MitigationPlan, r.OwnerId, r.DueDate, r.ClosedAt,
        r.CreatedBy, r.CreatedAt, r.UpdatedAt, RiskRules.RiskLevel(r.Score));
}

File: backend/app/services/risk_service.py
"""Сервис реестра рисков: CRUD с аудитом и портфельная сводка (матрица 3×3).

Балл (score) хранится денормализованно и пересчитывается здесь при изменении
вероятности/влияния. Портфель агрегируется set-based-запросами без N+1."""

from __future__ import annotations

import uuid
from datetime import UTC, datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import NotFoundError
from app.models.enums import (
    ACTIVE_RISK_STATUSES,
    AuditAction,
    RiskCategory,
    RiskStatus,
    Role,
)
from app.models.risk import Risk
from app.models.user import User
from app.repositories.project_repo import ProjectRepository
from app.schemas.risk import RiskCreate, RiskUpdate
from app.services.audit_service import AuditService
from app.services.notification_service import NotificationService
from app.services.risk_rules import HIGH_THRESHOLD, compute_score, risk_level

# Внутренние роли — кому осмысленно слать уведомление о назначении владельцем.
_INTERNAL_ROLES = {Role.ADMIN, Role.DIRECTOR, Role.PM, Role.PRESALE, Role.ENGINEER, Role.FINANCE}


class RiskService:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session
        self.audit = AuditService(session)
        self.notifier = NotificationService(session)
        self.projects = ProjectRepository(session)

    async def list_for_project(
        self, project_id: uuid.UUID, *, status: RiskStatus | None = None
    ) -> list[Risk]:
        stmt = select(Risk).where(
            Risk.project_id == project_id, Risk.deleted_at.is_(None)
        )
        if status is not None:
            stmt = stmt.where(Risk.status == status)
        # Сначала активные и более критичные, затем по дате.
        stmt = stmt.order_by(Risk.score.desc(), Risk.created_at.desc())
        return list((await self.session.execute(stmt)).scalars().all())

    async def get(self, risk_id: uuid.UUID) -> Risk:
        risk = await self.session.get(Risk, risk_id)
        if risk is None or risk.deleted_at is not None:
            raise NotFoundError("Риск не найден")
        return risk

    async def create(
        self, project_id: uuid.UUID, data: RiskCreate, actor_id: uuid.UUID
    ) -> Risk:
        score = compute_score(data.probability, data.impact)
        risk = Risk(
            project_id=project_id,
            created_by=actor_id,
            score=score,
            **data.model_dump(),
        )
        self.session.add(risk)
        await self.session.flush()
        await self.audit.record(
            entity="Risk",
            entity_id=risk.id,
            action=AuditAction.CREATE,
            actor_id=actor_id,
            diff={"title": risk.title, "score": score, "level": risk_level(score)},
        )
        if risk.owner_id and risk.owner_id != actor_id:
            await self._notify_owner(risk)
        return risk

    async def update(
        self, risk_id: uuid.UUID, data: RiskUpdate, actor_id: uuid.UUID
    ) -> Risk:
        risk = await self.get(risk_id)
        changes = data.model_dump(exclude_unset=True)
        prev_owner = risk.owner_id
        diff: dict = {}
        for field, value in changes.items():
            old = getattr(risk, field)
            if old != value:
                diff[field] = {"from": str(old), "to": str(value)}
                setattr(risk, field, value)

        # Пересчёт балла при изменении вероятности/влияния.
        if "probability" in changes or "impact" in changes:
            new_score = compute_score(risk.probability, risk.impact)
            if new_score != risk.score:
                diff["score"] = {"from": str(risk.score), "to": str(new_score)}
                risk.score = new_score

        # Метка закрытия: проставляется при переходе в closed, снимается при возврате.
        if "status" in changes:
            if risk.status == RiskStatus.CLOSED and risk.closed_at is None:
                risk.closed_at = datetime.now(UTC)
            elif risk.status != RiskStatus.CLOSED:
                risk.closed_at = None

        if diff:
            await self.audit.record(
                entity="Risk",
                entity_id=risk.id,
                action=AuditAction.UPDATE,
                actor_id=actor_id,
                diff=diff,
            )
        await self.session.flush()

        # Уведомить нового владельца, если сменился.
        if risk.owner_id and risk.owner_id != prev_owner and risk.owner_id != actor_id:
            await self._notify_owner(risk)
        return risk

    async def delete(self, risk_id: uuid.UUID, actor_id: uuid.UUID) -> None:
        risk = await self.get(risk_id)
        risk.deleted_at = datetime.now(UTC)
        await self.audit.record(
            entity="Risk",
            entity_id=risk.id,
            action=AuditAction.DELETE,
            actor_id=actor_id,
            diff={"title": risk.title},
        )
        await self.session.flush()

    async def _notify_owner(self, risk: Risk) -> None:
        if risk.owner_id is None:
            return
        owner = await self.session.get(User, risk.owner_id)
        if owner is None or not (_INTERNAL_ROLES & set(owner.roles)):
            return
        await self.notifier.notify(
            risk.owner_id,
            kind="risk_assigned",
            title="Вы назначены ответственным за риск",
            body=risk.title,
            link=f"/projects/{risk.project_id}",
        )

    # ---------- Портфель рисков (set-based) ----------
    async def portfolio(self, user: User) -> dict:
        # Доступные пользователю проекты (РП — свои; admin/director/finance — все).
        projects, _ = await self.projects.list_for_user(user, limit=500)
        by_id = {p.id: p for p in projects}
        ids = list(by_id)
        if not ids:
            return _empty_portfolio()

        stmt = select(
            Risk.project_id, Risk.probability, Risk.impact, Risk.score, Risk.category
        ).where(
            Risk.project_id.in_(ids),
            Risk.status.in_(ACTIVE_RISK_STATUSES),
            Risk.deleted_at.is_(None),
        )
        rows_raw = (await self.session.execute(stmt)).all()

        # Агрегаты.
        matrix: dict[tuple[int, int], int] = {}
        by_category: dict[RiskCategory, int] = {}
        per_project: dict[uuid.UUID, dict[str, int]] = {}
        total_high = 0
        for project_id, probability, impact, score, category in rows_raw:
            matrix[(probability, impact)] = matrix.get((probability, impact), 0) + 1
            cat = RiskCategory(category)
            by_category[cat] = by_category.get(cat, 0) + 1
            agg = per_project.setdefault(
                project_id, {"active_count": 0, "high_count": 0, "top_score": 0}
            )
            agg["active_count"] += 1
            agg["top_score"] = max(agg["top_score"], score)
            if score >= HIGH_THRESHOLD:
                agg["high_count"] += 1
                total_high += 1

        matrix_cells = [
            {
                "probability": p,
                "impact": i,
                "score": p * i,
                "level": risk_level(p * i),
                "count": matrix.get((p, i), 0),
            }
            for p in range(1, 4)
            for i in range(1, 4)
        ]
        rows = [
            {
                "project_id": pid,
                "code": by_id[pid].code,
                "name": by_id[pid].name,
                "stage": by_id[pid].stage,
                "status": by_id[pid].status,
                "manager_id": by_id[pid].manager_id,
                **agg,
            }
            for pid, agg in per_project.items()
            if pid in by_id
        ]
        rows.sort(key=lambda r: (r["top_score"], r["active_count"]), reverse=True)
        return {
            "projects_count": len(rows),
            "total_active": len(rows_raw),
            "total_high": total_high,
            "matrix": matrix_cells,
            "by_category": [
                {"category": c, "count": n}
                for c, n in sorted(by_category.items(), key=lambda kv: kv[1], reverse=True)
            ],
            "rows": rows,
        }

    @staticmethod
    def can_see_portfolio(user: User) -> bool:
        return bool(
            {Role.ADMIN, Role.DIRECTOR, Role.FINANCE, Role.PM} & set(user.roles)
        )


def _empty_portfolio() -> dict:
    return {
        "projects_count": 0,
        "total_active": 0,
        "total_high": 0,
        "matrix": [
            {
                "probability": p,
                "impact": i,
                "score": p * i,
                "level": risk_level(p * i),
                "count": 0,
            }
            for p in range(1, 4)
            for i in range(1, 4)
        ],
        "by_category": [],
        "rows": [],
    }
